core/integer/div_spec.rb

```diff
@@ -32,35 +30,71 @@
       lambda { -10.div(0.0) }.should raise_error(ZeroDivisionError)
     end
 
-    it "raises a ZeroDivisionError when the given argument is 0" do
+    it "raises a ZeroDivisionError when the given argument is 0 and not a Float" do
       lambda { 13.div(0) }.should raise_error(ZeroDivisionError)
+      lambda { 13.div(-0) }.should raise_error(ZeroDivisionError)
     end
 
-    it "raises a TypeError when given a non-Integer" do
-      lambda {
-        (obj = mock('10')).should_receive(:to_int).any_number_of_times.and_return(10)
-        13.div(obj)
-      }.should raise_error(TypeError)
+    it "raises a TypeError when given a non-numeric argument" do
+      lambda { 13.div(mock('10')) }.should raise_error(TypeError)
       lambda { 5.div("2") }.should raise_error(TypeError)
+      lambda { 5.div(:"2") }.should raise_error(TypeError)
+      lambda { 5.div([]) }.should raise_error(TypeError)
     end
   end
 
   context "bignum" do
-    it_behaves_like(:bignum_divide, :div)
+    before :each do
+      @bignum = bignum_value(88)
+    end
+
+    it "returns self divided by other" do
+      @bignum.div(4).should == 2305843009213693974
+      @bignum.div(Rational(4, 1)).should == 2305843009213693974
+      @bignum.div(bignum_value(2)).should == 1
+
+      (-(10**50)).div(-(10**40 + 1)).should == 9999999999
+      (10**50).div(10**40 + 1).should == 9999999999
+
+      (-10**50).div(10**40 + 1).should == -10000000000
+      (10**50).div(-(10**40 + 1)).should == -10000000000
+    end
+
+    it "may be not accurate with Float argument" do
+      @bignum.div(1  ).should == 9223372036854775896
+      @bignum.div(1.0).should == 9223372036854775808
+
+      @bignum.div(4  ).should == 2305843009213693974
+      @bignum.div(4.0).should == 2305843009213693952
+
+      @bignum.div(21  ).should == 439208192231179804
+      @bignum.div(21.0).should == 439208192231179776
+    end
```

Comment:
looks like a bug

Reply:
I'm not sure, the Numeric#div rdoc says:
```
Implementation from Numeric
------------------------------------------------------------------------------
  num.div(numeric)  ->  integer

------------------------------------------------------------------------------

Uses / to perform division, then converts the result to an integer.
Numeric does not define the / operator; this is left to subclasses.
```

And if indeed it goes through Float then it loses precision:
```ruby
> 9223372036854775896/21.0
=> 4.392081922311798e+17
```

And the Integer#div rdoc is fairly imprecise:
```
Implementation from Integer
------------------------------------------------------------------------------
  int.div(numeric)  ->  integer

------------------------------------------------------------------------------

Performs integer division: returns the integer result of dividing int
by numeric.
```

The implementation is pretty simple:
https://github.com/ruby/ruby/blob/3a887648fe75c022c3f445e5943525226f15e3d0/numeric.c#L571-L588
It calls `/` then `floor`.
That would be a good spec and clarify the behavior.

Might still be worth asking on MRI's tracker for clarification though.